### ISSUE TITLE
Unthreaded receipt is not returned in the next syncs

### ISSUE THREAD

**Issue by florianduros** [element-hq/synapse#17247]
### Description

I'm investigating cases of stuck unread on EW.
When I'm sending an unthreaded receipt on a room with unread threads, my receipt doesn't appear in the next syncs.
Maybe is due to the last event of the room is an user renameing

### Steps to reproduce

In a room:
- Unread threads
- Main timeline is read
- The last event of the room is an user renaming

I send an unthreaded receipt:

```
POST https://element.ems.host/_matrix/client/v3/rooms/!jSkNtsjcsXfCRJwGpp%3Amatrix.org/receipt/m.read/%24UVcMglbqEZkuH2fq5zzbH-gQGyIU3MKr83qdtWUJwHg
```
The next syncs don't include my receipt.

RoomId: !jSkNtsjcsXfCRJwGpp:matrix.org
Last important eventId:  $UVcMglbqEZkuH2fq5zzbH-gQGyIU3MKr83qdtWUJwHg

### Homeserver

element.io

### Synapse Version

1.107.0

### Installation Method

I don't know

### Database

.

### Workers

Single process

### Platform

.

### Relevant log output

```shell
.
```

**Comment by devonh** [comment]
When I try to reproduce this it seems to me like things are working as expected.
But maybe my expectations don't line up with what is actually expected.

So here's what I'm doing: (running synapse v1.107.0 locally, using app.element.io as a client)
- create a room as userA (private, unencrypted)
- invite userB
- join the room as userB
- userA sends a message
- userB sends a message
- userB sends a threaded reply to UserA's message
- userA changes their display name

This puts things in the following state:
- userA has read the main timeline
- userA has an unread notification on the Threads icon
- userB has read both the main timeline & the thread

If I now do a POST as mentioned above using the eventID for userA's display name change, the unread symbol on the Threads icon goes away for userA.

Now, if userA's last event was before the last event in the thread (ie. userA sends another display name change, then userB sends another message in the thread):
- When sending the POST using the event ID of userA's latest name change event, which is the latest event in the main timeline, the unread symbol on the Threads icon remains for userA
  - which I believe is due to the receipt referencing an event in the timeline that is older than the latest event in the room (ie. the newer thread event)

As far as I can tell, this is the desired behaviour.

**Comment by devonh** [comment]
So a couple of questions:
- Is the behaviour I described not the expected outcome?
- If the behaviour I described is the expected outcome, then is there any other information you can provide that might have been necessary to reproduce this issue? (ie. type of room, ordering of messages, etc.)

**Comment by clokep** [comment]
@devonh Sounds like you're correct to me, but I've found drawing out the DAGs (including relations) and labeling each node to be immensely beneficial in these situations.

**Comment by florianduros** [comment]
> When I try to reproduce this it seems to me like things are working as expected. But maybe my expectations don't line up with what is actually expected.
> 
> So here's what I'm doing: (running synapse v1.107.0 locally, using app.element.io as a client)
> 
> * create a room as userA (private, unencrypted)
> * invite userB
> * join the room as userB
> * userA sends a message
> * userB sends a message
> * userB sends a threaded reply to UserA's message
> * userA changes their display name
> 
> This puts things in the following state:
> 
> * userA has read the main timeline
> * userA has an unread notification on the Threads icon
> * userB has read both the main timeline & the thread
> 
> If I now do a POST as mentioned above using the eventID for userA's display name change, the unread symbol on the Threads icon goes away for userA.
> 
> Now, if userA's last event was before the last event in the thread (ie. userA sends another display name change, then userB sends another message in the thread):
> 
> * When sending the POST using the event ID of userA's latest name change event, which is the latest event in the main timeline, the unread symbol on the Threads icon remains for userA
>   
>   * which I believe is due to the receipt referencing an event in the timeline that is older than the latest event in the room (ie. the newer thread event)
> 
> As far as I can tell, this is the desired behaviour.

Yes this is the expected behaviour but it is not the behaviour I'm seeing every time when I'm using EW + element homeserver.

If I'm using your example, sometimes (I don't know what it is causing this behaviour) when userA send an unthreaded receipt to the last event Id, the next sync doesn't include our new receipt.
In EW, the userB thread is viewed as seen thanks to an EW internal mechanism but if EW is reloaded, the thread is unread again (because the receipt was missing in the sync).

A lot of unread spikes are happening due to this missing receipt in the sync.

I can provide real data, I always have some rooms suffering this behaviour.

**Comment by devonh** [comment]
If you could provide data/logs around the time where this is happening that could be very helpful.
I am failing to reproduce this issue with my local setup.

**Comment by kieranlane** [comment]
Hi @devonh @florianduros, just following up on this as trying to clear away old tickets before the move to JIRA, this appears related to this [support ticket](https://newvector.zammad.com/#ticket/zoom/56690) in Zammad. Looks like perhaps this is pending logs?

**Comment by florianduros** [comment]
Hi, I'll try to reproduce it next week and provide logs.
